src/components/Grid.js

```diff
@@ -66,11 +67,45 @@ export default class Grid extends Component {
           ]
 
           let {onDataChange = () => {}} = this.props
-          onDataChange({data: {nodes: newNodes, edges: newEdges}})
+          onDataChange({data: {nodes: newNodes, edges: newEdges}}, true)
         }
 
         this.setState({phantomArrow: null})
       }
+      if (nodesToJoin != null) {
+        let [draggedNode, existingNode] = nodesToJoin
+
+        if (draggedNode.id !== existingNode.id) {
+          let [textNode, emptyNode] =
+            draggedNode.value.length > 0
+              ? [draggedNode, existingNode]
+              : [existingNode, draggedNode]
+
+          let {onDataChange = () => {}} = this.props
+          onDataChange({
+            data: {
+              nodes: this.props.data.nodes
+                .filter(node => node.id != textNode.id)
+                .map(node =>
+                  node.id == emptyNode.id
+                    ? {
+                        ...emptyNode,
+                        position: existingNode.position,
+                        value: textNode.value
+                      }
+                    : node
+                ),
+              edges: this.props.data.edges.map(edge => {
+                if (edge.from == textNode.id) edge.from = emptyNode.id
+                if (edge.to == textNode.id) edge.to = emptyNode.id
+                if (edge.to == edge.from)
+                  edge = {...edge, loop: [0, false], labelPosition: 'right'}
```

Comment:
Why are we resetting the loop properties?

Reply:
It wasn't meant to reset the loop properties, but rather set them the first time, when the start of an arrow is dragged to its head. I forgot that the properties might already be set.